### PR TITLE
Improve cursor window format in `StringScanner#inspect`

### PR DESCRIPTION
**Pull request by jneen**
Depends on #16593 for `#peek_behind`.

Extracted from #16455.

## Rationale

StringScanner#inspect is currently very basic. It shows the character offset, and what looks like a string literal (but without proper escapes) of 2 characters behind and 3 characters ahead of the scan head. This not only obfuscates where the scan head is, it also uses `@str[index, count]`, which, as in #16593, can be linear-time in terms of the offset.

## Changes

This change, based partially on [discussion here](https://github.com/crystal-lang/crystal/pull/16455#discussion_r2606393806) with @straight-shoota:

* Uses byte offsets to eliminate the linear-time scans with `#peek` and `#peek_behind`
* Indicates the scan head using `‣`
* Shows `"` at the beginning and ending of the input, or `…` at the ending, giving a better indicator of where the scan head is in relation to the edges of the string
* Properly escapes special characters using `String#inspect_unquoted`

Some examples of the inspect output for single-byte and multi-byte strings, including examples with spaces at the boundaries:

```
#<StringScanner 0/16 "‣this …>
#<StringScanner 5/16 …s ‣is …>
#<StringScanner 6/16 … i‣s a…>
#<StringScanner 8/16 …s ‣a s…>
#<StringScanner 16/16 …tring‣">
#<StringScanner 0/2 "‣hi">
#<StringScanner 1/2 "h‣i">
#<StringScanner 2/2 "hi‣">
#<StringScanner 1/5 " ‣a\n\t ">
#<StringScanner 0/9 "‣これは文字…>
#<StringScanner 1/9 "こ‣れは文字…>
#<StringScanner 2/9 "これ‣は文字…>
#<StringScanner 3/9 …れは‣文字列…>
#<StringScanner 6/9 …字列‣である">
#<StringScanner 9/9 …字列である‣">
```

### CODE REVIEW

File: src/string_scanner.cr [hunk 1]
```diff
@@ -448,16 +448,60 @@ class StringScanner
...
+      chars_after = remaining
+
+      # no ellipsis if we hit the end
+      ellipsis_end = false
+    end
+
+    # clamp in the case of small strings that don't have enough
+    # before *or* after
+    chars_before = chars_before.clamp(0, offset)
+    chars_after = chars_after.clamp(0, remaining)
```
Comment (straight-shoota) [inline src/string_scanner.cr]: **suggestion:** I believe this boils down to two relatively straightforward clamping expression:
```suggestion
    chars_before = offset.clamp(0, (5 - remaining).clamp(2..))
    chars_after = remaining.clamp(0, 5 - chars_before)

    ellipsis_start = chars_before < offset
    ellipsis_end = chars_after < remaining
```

The two bool expressions could probably be inlined as well. The local vars are not really useful.

Comment (jneen) [reply src/string_scanner.cr]: This is certainly an elegant solution but I'm having trouble wrapping my head around it - perhaps a comment or two would be helpful?

Comment (straight-shoota) [reply src/string_scanner.cr]: Sure! Let's start with `chars_after` which is the simpler expression. At this point we already know how many characters were printing before the cursor. So after we print as many as available (`remaining`) but limited so that the total amount on both sides does not exceed 5.

`chars_before` is effectively in the range `0..5`.
We achieve that by basing it on the amount of available characters (`offset`), but at most 2, or more if `remaining` is small so that we wouldn't fill the 5 character window.

Does that make sense? It would be possible to rewrite the clamps to max/min expressions which might make it easier to comprehend, but less succinct.

Comment (jneen) [reply src/string_scanner.cr]: I see! I think I might prefer the min/max expressions in this case, because the extra clamp from zero doesn't seem necessary as long as we trust that offset and remaining are non-negative - and we have a lot of edge cases in the tests for things like being at the very beginning or end, or having a very small string. I'm also adding a spec for empty string, where both offset and remaining are 0, but it seems like neither implementation has any problem with it.

```crystal
    # find a range of 5 characters with the scan head as close to
    # the middle as possible.
    chars_after = Math.min(remaining, Math.max(3, 5 - offset))
    chars_before = Math.min(offset, 5 - chars_after)
```

(Also realizing `Comparable#min(other)` and `Comparable#max(other)` could be convenient)

Comment (jneen) [reply src/string_scanner.cr]: Or maybe even better, putting the usual expected behaviour out in front and letting the final clamp expression be a clear range check:

```crystal
    chars_after = Math.max(3, 5 - offset).clamp(0, remaining)
    chars_before = (5 - chars_after).clamp(0, offset)
```

**Comment by jneen** [comment]
Apologies for the push, was just reviewing the docs and found these private constants showed up in the documentation.

**Comment by Blacksmoke16** [comment]
If they truly are private to the string scanner type, maybe best to do like `private INSPECT_INDICATOR_CHAR = ...` instead of just nodocing them?

**Comment by jneen** [comment]
Legit didn't know constants could be private. That's pretty cool!

**Comment by ysbaddaden** [comment]
This is very nice :heart_eyes:
